File: week-2/problem-set/prob8.py
"""Prob8 example.

This file belongs to the week 2 data-structure and graph practice solutions. It uses Queue to illustrate the main idea with a small runnable example.
"""

class Queue:
    """Queue implemented with two stacks."""
    def __init__(self):
        """Run the   init   step used in this example.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.enqueue_stack = []
        self.dequeue_stack = []

    def enqueue(self, item):
        """Insert a new item into the data structure.

        Args:
            item (object): The item input used by this function.

        Returns:
            None: This function updates data or prints results and does not return a value.
        """
        self.enqueue_stack.append(item)

    def dequeue(self):
        """Remove and return the next item from the data structure.

        Returns:
            object: The value produced by this function.
        """
        if not self.dequeue_stack:
            while self.enqueue_stack:
                self.dequeue_stack.append(self.enqueue_stack.pop())
        if not self.dequeue_stack:
            raise IndexError("Queue is empty")
        return self.dequeue_stack.pop()

    def is_empty(self):
        """Run the is empty step used in this example.

        Returns:
            bool: True when the required condition is satisfied, otherwise False.
        """
        return len(self.enqueue_stack) == 0 and len(self.dequeue_stack) == 0

q = Queue()
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
print(q.dequeue()) # Output: 1
q.enqueue(4)
print(q.dequeue()) # Output: 2
print(q.dequeue()) # Output: 3
print(q.dequeue()) # Output: 4
print(q.is_empty()) # Output: True

File: week-2/problem-set/prob8.cpp
/**
 * @file prob8.cpp
 * @brief C++ version of the queue-using-two-stacks example.
 */

#include <iostream>
#include <stack>

using namespace std;

class Queue {
public:
    /**
     * @brief Insert a new item at the back of the queue.
     * @param item int Value to add to the queue.
     * @return void This method updates the queue and does not return a value.
     */
    void enqueue(int item) {
        enqueue_stack.push(item);
    }

    /**
     * @brief Remove and return the item at the front of the queue.
     * @return int Front value removed from the queue.
     */
    int dequeue() {
        if (dequeue_stack.empty()) {
            while (!enqueue_stack.empty()) {
                dequeue_stack.push(enqueue_stack.top());
                enqueue_stack.pop();
            }
        }
        if (dequeue_stack.empty()) {
            throw out_of_range("Queue is empty");
        }
        int value = dequeue_stack.top();
        dequeue_stack.pop();
        return value;
    }

    /**
     * @brief Return true when the queue stores no items.
     * @return bool True when both internal stacks are empty, otherwise false.
     */
    bool is_empty() const {
        return enqueue_stack.empty() && dequeue_stack.empty();
    }

private:
    stack<int> enqueue_stack;
    stack<int> dequeue_stack;
};

/**
 * @brief Run the queue demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    Queue q;
    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);
    cout << q.dequeue() << '\n';
    q.enqueue(4);
    cout << q.dequeue() << '\n';
    cout << q.dequeue() << '\n';
    cout << q.dequeue() << '\n';
    cout << boolalpha << q.is_empty() << '\n';
    return 0;
}
